File: backend/src/main/java/com/sliit/smartcampus/modules/booking/service/BookingService.java
package com.sliit.smartcampus.modules.booking.service;

import com.sliit.smartcampus.modules.booking.model.Booking;
import com.sliit.smartcampus.modules.booking.repository.BookingRepository;
import com.sliit.smartcampus.common.enums.BookingStatus;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class BookingService {

    private final BookingRepository bookingRepository;

    // Retrieve all bookings (for Admin view)
    public List<Booking> getAllBookings() {
        return bookingRepository.findAll();
    }

    // Retrieve bookings for a specific user
    public List<Booking> getBookingsByUser(String userEmail) {
        return bookingRepository.findByUserEmail(userEmail);
    }

    // Create a new booking with Conflict Checking
    public Booking createBooking(Booking booking) {
        // CONFLICT CHECK: Are there any overlapping bookings?
        List<Booking> conflicts = bookingRepository.findConflictingBookings(
                booking.getFacilityId(),
                booking.getBookingDate(),
                booking.getStartTime(),
                booking.getEndTime()
        );

        if (!conflicts.isEmpty()) {
            throw new RuntimeException("Scheduling conflict: The resource is already booked for the selected time range.");
        }

        // Set initial status
        booking.setStatus(BookingStatus.PENDING);
        return bookingRepository.save(booking);
    }

    // Admin updates the status (Approve or Reject)
    public Booking updateBookingStatus(String bookingId, BookingStatus newStatus, String adminReason) {
        Booking booking = bookingRepository.findById(bookingId)
                .orElseThrow(() -> new RuntimeException("Booking not found"));

        booking.setStatus(newStatus);

        // If rejected, an admin reason should be provided
        if (newStatus == BookingStatus.REJECTED && adminReason != null) {
            booking.setAdminReason(adminReason);
        }

        return bookingRepository.save(booking);
    }

    // Delete a booking
    public void deleteBooking(String bookingId) {
        bookingRepository.deleteById(bookingId);
    }
}

File: frontend/src/services/bookingService.js
import api from './api';

const bookingService = {
    // 1. Create a new booking
    createBooking: async (bookingData) => {
        try {
            const response = await api.post('/bookings', bookingData);
            return response.data;
        } catch (error) {
            if (error.response && error.response.data && error.response.data.error) {
                throw new Error(error.response.data.error);
            }
            throw new Error("Failed to create booking.");
        }
    },

    // 2. Fetch bookings for a specific user
    getUserBookings: async (email) => {
        const response = await api.get(`/bookings?userEmail=${email}`);
        return response.data;
    },

    // 3. Fetch ALL bookings (For Admin)
    getAllBookings: async () => {
        const response = await api.get('/bookings');
        return response.data;
    },

    // 4. Update Booking Status (Approve/Reject)
    updateBookingStatus: async (id, status, adminReason = "") => {
        const response = await api.put(`/bookings/${id}/status`, { status, adminReason });
        return response.data;
    },

    // 5. Delete/Cancel a booking
    deleteBooking: async (id) => {
        const response = await api.delete(`/bookings/${id}`);
        return response.data;
    }
};

export default bookingService;